### PR TITLE
Implement `log2`, `ln`, `log10`

### LINKED ISSUE #84: [Feature]: `ln` for fixed-point math lib
### What is the feature you would like to see?

Implement natural logarithm ln(x).

### LINKED ISSUE #85: [Feature]: `log10` for fixed-point math lib
### What is the feature you would like to see?

Implement common logarithm log10(x).

### LINKED ISSUE #86: [Feature]: `log2` for fixed-point math
### What is the feature you would like to see?

Implement binary logarithm log2(x).

### PR DESCRIPTION
**Pull request by immrsd**
### Summary

Add `log2`, `ln`, `log10` to `UD30x9` and `SD29x9`
  
Adds the three logarithm functions to both fixed-point types, completing the transcendental kernel
started by sqrt.

### Algorithm

raw_log2 (in internal/common.move) runs a bit-by-bit squaring loop at internal precision 10^18 and
returns a (neg, magnitude_at_10^18) tuple shared by all six public functions. The 10^18 internal
scale (vs. the user-facing 10^9) gives the loop enough headroom that inputs near 1.0 — where logs
shrink to a handful of ulps and any internal error would dominate — still produce correctly-rounded
UD30x9 results. ln and log10 are thin wrappers over apply_log2_factor, which multiplies the log2
magnitude by a base-conversion constant (ln2_e18 or log10_2_e18) and downscales to 10^9.

### Behavior

- UD30x9: rounds down (consistent with sqrt). Aborts on x < 1 with ELogUndefined.
- SD29x9: rounds toward zero (consistent with mul_trunc, div_trunc, pow in the same module). Aborts
on x <= 0 with ELogUndefined.

## Release Notes

* **New Features**
  * Added natural logarithm, base-10 logarithm, and base-2 logarithm operations to fixed-point math libraries for both signed and unsigned numeric types.

* **Tests**
  * Comprehensive test coverage including boundary conditions, reference value validation, and property-based testing for logarithm functions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/OpenZeppelin/contracts-sui/pull/320)

### CODE REVIEW

File: math/fixed_point/sources/internal/common.move [hunk 1]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
 }
+
+/// Internal precision used by the fixed-point logarithm kernel: `10^18`
+/// (~`2^60`), an order of magnitude finer than the user-facing `10^9` scale.
+///
+/// Error analysis: in the squaring loop, error in `y` grows ~2× per iteration
+/// (since `y_new = y_old^2 / internal`), but a wrong bit at iteration `i`
+/// only perturbs `frac` by `internal / 2^(i+1)` — exponentially decaying, so
+/// late-iteration errors contribute little. Empirically the total `frac`
+/// error stays under `~10^3` at scale `10^18` (verified by `raw_log2_tests`),
```
Comment (0xNeshi) [inline math/fixed_point/sources/internal/common.move]:
```suggestion
/// error stays under `~10^3` at scale `10^18`,
```
nit: fragile + redundant

File: math/fixed_point/sources/internal/common.move [hunk 2]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+/// well below one user-facing ulp (`10^9`). A 9-decimal variant would lack
```
Comment (0xNeshi) [inline math/fixed_point/sources/internal/common.move]:
```suggestion
/// well below one user-facing up (`10^9`). A 9-decimal variant would lack
```
typo?

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 3]
```diff
@@ -201,6 +205,74 @@ public fun is_zero(x: SD29x9): bool {
...
+/// see `log2` for full rounding semantics on signed results.
+///
+/// #### Parameters
+/// - `x`: Input value. Must be strictly positive.
+///
+/// #### Returns
+/// - `ln(x)`, rounded toward zero.
+///
+/// #### Aborts
+/// - `ELogUndefined` if `x` is zero or negative.
```
Comment (0xNeshi) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move]:
```suggestion
/// - `x`: Input value.
///
/// #### Returns
/// - `ln(x)`, rounded toward zero.
///
/// #### Aborts
/// - `ELogUndefined` if `x` is zero or negative.
```
nit: potentially redundant; already documented in `Aborts` section

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 4]
```diff
@@ -201,6 +205,74 @@ public fun is_zero(x: SD29x9): bool {
...
+    wrap_components(Components { neg: log_neg, mag: result_mag as u256 })
+}
+
+/// Computes the base-10 logarithm of an `SD29x9` value.
+///
+/// Derived from `log2` via `log10(x) = log2(x) * log10(2)`. Rounded toward
+/// zero; see `log2` for full rounding semantics on signed results.
+///
+/// #### Parameters
+/// - `x`: Input value. Must be strictly positive.
```
Comment (0xNeshi) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move]:
```suggestion
/// - `x`: Input value.
```
nit: same as above

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 5]
```diff
@@ -201,6 +205,74 @@ public fun is_zero(x: SD29x9): bool {
...
+}
+
+/// Computes the base-2 logarithm of an `SD29x9` value.
+///
+/// The result is rounded toward zero, matching the convention used by
+/// `mul_trunc`, `div_trunc`, and `pow` in this module. For positive results
+/// (inputs `>= 1`) this coincides with rounding down. For negative results
+/// (inputs in `(0, 1)`) the signed result usually sits closer to zero than
+/// the true value, but in narrow edge cases where the kernel's small upward
+/// magnitude bias crosses an integer boundary it may instead be 1 ulp
```
Comment (0xNeshi) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move]: What's "ulp"? Devs reading this might not know either, consider either documenting what it means or using a different term

File: math/fixed_point/sources/sd29x9/sd29x9_base.move [hunk 6]
```diff
@@ -201,6 +205,74 @@ public fun is_zero(x: SD29x9): bool {
...
+/// further from zero.
+///
+/// #### Parameters
+/// - `x`: Input value. Must be strictly positive.
```
Comment (0xNeshi) [inline math/fixed_point/sources/sd29x9/sd29x9_base.move]:
```suggestion
/// - `x`: Input value.
```
nit: if you don't think it's redundant, feel free to ignore

File: math/fixed_point/sources/ud30x9/ud30x9_base.move [hunk 7]
```diff
@@ -243,6 +248,74 @@ public fun unchecked_lshift(x: UD30x9, bits: u8): UD30x9 {
     wrap(x.unwrap() << bits)
 }
 
+/// Computes the natural logarithm of a `UD30x9` value.
+///
+/// Derived from `log2` via the identity `ln(x) = log2(x) * ln(2)`. Both factors
+/// round toward zero, so the result may sit up to one ulp below the true value.
+///
+/// #### Parameters
+/// - `x`: Input value. Must be greater than or equal to one.
```
Comment (0xNeshi) [inline math/fixed_point/sources/ud30x9/ud30x9_base.move]:
```suggestion
/// - `x`: Input value.
```

**Review by 0xNeshi — COMMENTED** [review]
First iteration, mostly nits

File: math/fixed_point/README.md [hunk 8]
```diff
@@ -86,6 +87,31 @@ Because Move does not provide a native signed integer type for this package,
...
+
+- **UD30x9** aborts on `x < 1` (the result would be negative).
+  Rounds down.
+- **SD29x9** aborts on `x <= 0`. Rounds toward zero, matching `mul_trunc`,
+  `div_trunc`, and `pow` in the same module.
+
+Round-down compounding can put the result 1 ulp from the mathematical answer
+at irrational identity points (e.g. `log10(10·SCALE) == SCALE - 1`).
+
+```rust
```
Comment (bidzyyys) [inline math/fixed_point/README.md]: Shouldn't be `move`?; Applies to all occurrences.

File: math/fixed_point/sources/internal/common.move [hunk 9]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+/// the headroom for inputs near `1.0`; 18-decimal preserves precision.
+///
+/// #### Returns
+/// - `10^18` as `u128`.
+public(package) macro fun internal_log_scale(): u128 {
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move]: Can't it be just a const value?

File: math/fixed_point/sources/internal/common.move [hunk 10]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+    1_000_000_000_000_000_000 // 10^18
+}
+
+/// `ln(2)` represented at scale `10^18`, rounded down.
+///
+/// #### Returns
+/// - `floor(ln(2) * 10^18) = 693_147_180_559_945_309`.
+public(package) macro fun ln2_e18(): u128 {
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move:104]: Same as above

File: math/fixed_point/sources/internal/common.move [hunk 11]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+    693_147_180_559_945_309
+}
+
+/// `log10(2)` represented at scale `10^18`, rounded down.
+///
+/// #### Returns
+/// - `floor(log10(2) * 10^18) = 301_029_995_663_981_195`.
+public(package) macro fun log10_2_e18(): u128 {
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move:112]: Same as above

File: math/fixed_point/sources/internal/common.move [hunk 12]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+///   raw form.
+public(package) fun apply_log2_factor(log2_mag_e18: u128, factor_e18: u128): u128 {
+    // Two scale-10^18 factors yield a scale-10^36 product; dividing by 10^27
+    // lands the result at scale 10^9 = UD30x9/SD29x9 raw. `log2_mag_e18 < 2^67`
+    // and `factor_e18 < 2^60`, so the product reaches up to ~2^127 — right at
+    // the `u128` boundary. `u128::mul_div` widens the product to `u256`
+    // internally before dividing, so the intermediate value never overflows;
+    // the final quotient (after `/ 10^27`) safely fits back in `u128`.
+    let denom: u128 = 1_000_000_000_000_000_000_000_000_000; // 10^27
+    u128::mul_div(log2_mag_e18, factor_e18, denom, rounding::down()).destroy_some()
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move]: Can't we have `log2_mag_e18.mul_div(factor_e18, denom, rounding::down()).destroy_some()`?

File: math/fixed_point/sources/internal/common.move [hunk 13]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+/// - The magnitude at scale `10^9`, ready to wrap into `UD30x9` or `SD29x9`
+///   raw form.
+public(package) fun apply_log2_factor(log2_mag_e18: u128, factor_e18: u128): u128 {
+    // Two scale-10^18 factors yield a scale-10^36 product; dividing by 10^27
+    // lands the result at scale 10^9 = UD30x9/SD29x9 raw. `log2_mag_e18 < 2^67`
+    // and `factor_e18 < 2^60`, so the product reaches up to ~2^127 — right at
+    // the `u128` boundary. `u128::mul_div` widens the product to `u256`
+    // internally before dividing, so the intermediate value never overflows;
+    // the final quotient (after `/ 10^27`) safely fits back in `u128`.
+    let denom: u128 = 1_000_000_000_000_000_000_000_000_000; // 10^27
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move]: Const value?

File: math/fixed_point/sources/internal/common.move [hunk 14]
```diff
@@ -87,3 +96,129 @@ public(package) fun div_away_u256(numerator: u256, denominator: u256): u256 {
...
+
+    // Lift from scale `10^9` to scale `10^18` for the iteration. After lift
+    // `y < 2 * 10^18 < 2^61`, so `y * y < 2^122` fits in `u128`.
+    let mut y: u128 = y_at_scale * scale;
+    let internal_x2: u128 = 2 * internal;
+
+    let mut frac: u128 = 0;
+    let mut delta: u128 = internal / 2;
+    while (delta > 0) {
+        y = y * y / internal;
```
Comment (bidzyyys) [inline math/fixed_point/sources/internal/common.move:210]: `mul_div`?

**Review by bidzyyys — COMMENTED** [review]
First pass.

File: math/fixed_point/sources/internal/common.move (hunk 2 above)
Comment (0xNeshi) [reply math/fixed_point/sources/internal/common.move]: Is `ulp` = https://en.wikipedia.org/wiki/Unit_in_the_last_place ?

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 5 above)
Comment (0xNeshi) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move]: Is it this https://en.wikipedia.org/wiki/Unit_in_the_last_place ?

File: math/fixed_point/sources/internal/common.move (hunk 10 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move:104]: Kept as a macro because `ln2_e18` is read from `sd29x9_base` and `ud30x9_base` as well — `const` in Move is module-local, so `public(package) macro` is the simplest way to share the literal across the package.

File: math/fixed_point/sources/internal/common.move (hunk 11 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move:112]: Same rationale as `ln2_e18` — used from both `sd29x9_base` and `ud30x9_base`, so kept as a `public(package) macro` rather than a module-local `const`.

File: math/fixed_point/sources/internal/common.move (hunk 14 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move:210]: `y * y` is safe here: the loop preserves the invariant `y < 2 * internal = 2·10^18 < 2^61` (restored each iteration by the `y >> 1` branch when `y >= 2 * internal`), so `y * y < 2^122` always fits in `u128`. Native `*` / `/` avoids `mul_div`'s `u256` widen + `Option` unwrap per iteration and gives the same round-down result. Extended the comment above the loop to spell out the inductive invariant.

File: math/fixed_point/sources/internal/common.move (hunk 13 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move]: Extracted to a const

File: math/fixed_point/sources/internal/common.move (hunk 12 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move]: Applied

File: math/fixed_point/sources/internal/common.move (hunk 9 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move]: Made a constant

File: math/fixed_point/README.md (hunk 8 above)
Comment (immrsd) [reply math/fixed_point/README.md]: Applied

File: math/fixed_point/sources/ud30x9/ud30x9_base.move (hunk 7 above)
Comment (immrsd) [reply math/fixed_point/sources/ud30x9/ud30x9_base.move]: Applied

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 6 above)
Comment (immrsd) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move]: Applied

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 4 above)
Comment (immrsd) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move]: Applied

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 3 above)
Comment (immrsd) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move]: Applied

File: math/fixed_point/sources/internal/common.move (hunk 1 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move]: Applied

File: math/fixed_point/tests/sd29x9_tests/log2_tests.move [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+#[test]
+fun log2_of_one_third_pins_value() {
+    // x_raw = 333_333_333 represents 0.333333333 (one raw ulp below true 1/3).
+    // True log2(0.333333333) ≈ -1.5849625021638512. floor of |that| at 10^9
+    // scale is 1_584_962_502; the algorithm returns this magnitude with the
+    // negative sign.
+    let result = pos(333_333_333).log2();
+    assert_eq!(result.abs().unwrap(), 1_584_962_502);
+}
```
[the remark this answers is not in the document]
Comment (immrsd) [reply math/fixed_point/tests/sd29x9_tests/log2_tests.move:57]: Applied

File: math/fixed_point/sources/internal/common.move (hunk 2 above)
Comment (immrsd) [reply math/fixed_point/sources/internal/common.move]: Yes, it'd indeed "unit in the last price". Added clarification for the 1st usage of the term in the file

File: math/fixed_point/sources/sd29x9/sd29x9_base.move (hunk 5 above)
Comment (immrsd) [reply math/fixed_point/sources/sd29x9/sd29x9_base.move]: Added clarification

**Review by bidzyyys — APPROVED** [review]
LGTM

**Review by 0xNeshi — APPROVED** [review]
